pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,664 @@
+            dep.find("/") > 0 and
+            dep.find("://") < 0 and
+            (dep.find("@") < 0 or
+             dep.find("@") > dep.find("/"))):
+        return True
+    return False
+
+
+def parse_as_folder_reference(dep):
+    """ See if a dependency reference refers to a folder path.
+        If it does, return the folder path (which parses and
+        resolves file:// urls in the process).
+        If it doesn't, return None.
+    """
+    if dep.startswith("/") or dep.startswith("file://") or (
+            dep.find("/") > 0 and
+            dep.find("://") < 0):
+        if dep.startswith("file://"):
+            dep = urlparse(dep).path
+        return dep
+    return None
+
+
+def _extract_info_from_package(dependency, extract_type=None, debug=False):
+    """ Internal function to extract metainfo from a package.
+        Currently supported info types:
+
+        - name
+        - namewithpins  (an exact version pin is kept if any)
+        - dependencies  (a list of dependencies)
+    """
+    output_folder = tempfile.mkdtemp(prefix="pythonpackage-metafolder-")
+    try:
+        extract_metainfo_files_from_package(
+            dependency, output_folder, debug=debug
+        )
+
+        with open(os.path.join(output_folder, "METADATA"),
+                  "r", encoding="utf-8"
+                 ) as f:
+            # Get metadata and cut away description (is after 2 linebreaks)
+            metadata_entries = f.read().partition("\n\n")[0].splitlines()
+
+        if extract_type == "name" or extract_type == "namewithpins":
+            name = None
+            for meta_entry in metadata_entries:
+                if meta_entry.lower().startswith("name:"):
+                    return meta_entry.partition(":")[2].strip()
+            if name is None:
+                raise ValueError("failed to obtain package name")
+
+            if extract_type == "namewithpins" and \
+                    dependency.find("==") > 0 and \
+                    dependency.partition("==")[0].lower() == name.lower():
+                return name + "==" + dependency.partition("==")[2].lstrip()
+            return name
+        elif extract_type == "dependencies":
+            requirements = []
+            if os.path.exists(os.path.join(output_folder, 'pyproject.toml')):
+                with open(os.path.join(output_folder, 'pyproject.toml')) as f:
+                    build_sys = pytoml.load(f)['build-system']
+                    if "requires" in build_sys:
+                        requirements += build_sys["requires"]
+
+            # Add requirements from metadata:
+            requirements += [
+                entry.rpartition("Requires-Dist:")[2].strip()
+                for entry in metadata_entries
+                if entry.startswith("Requires-Dist")
+            ]
+
+            return list(set(requirements))  # remove duplicates
+    finally:
+        shutil.rmtree(output_folder)
+
+
+package_name_cache = dict()
```

Comment:
is this all that slow that we want to implement our own caching?

Reply:
@AndreMiras fun fact, to determine the name of a python package you need to install everything in its `setup_requires` or `build_requires`. So yes, for my project this actually runs like a minute long :nauseated_face:  since the name is queried multiple times I added this cache

Reply:
(I discussed this with pip / setuptools folks btw, it really just is like that, inherently. so it's not something we're doing wrong, there is no other way due to how much code injection python packaging scripts allow)

Reply:
resolving for now since I don't see a better way, but feel free to poke/unresolve if you still got some ideas on this :+1: